### ISSUE TITLE
Proposal for an SC – Data input limits

### ISSUE THREAD

**Issue by guyhickling** [w3c/wcag21#329]
This Success Criterion is proposed to counter the quite common problem of developers not bothering to include basic attributes that should limit what users can enter into data fields. 

HTML, for instance, provides the “maxlength” attribute for text input fields. This stops users from adding input once they reach the maximum length it specifies. And HTML5 provides maximum and minimum attributes to limit numeric input.

It can be extremely annoying for users who have typed 200 characters into a text field, for example, to then be told on form submission that their effort was wasted and they must now go back to edit what they typed because the maximum allowed is 150 characters! But having to backtrack and edit is much more difficult for some groups of disabled people who may already have difficulty completing forms. 

This has no impact on any other criteria proposed for 2.1 (so far as I can see), or anything in 2.0. It simply fills in a small hole that has not been touched so far.

# SC Shortname
Data Input Limits

# SC Text
For all data input fields that accept character input the following are true where the technology used has provision for it:
-	for input of text: the maximum number of characters of input allowed is specified in the field markup
-	for input of numeric values: the maximum allowed value is specified and, if the minimum allowed value is not zero, the minimum allowed value is specified in the field markup,
-	the above specified maximum number of characters, or maximum allowed value, does not exceed the length or value (and minimum specified does not fall below the value) that can be handled by subsequent processing including storing to a database.

Input fields that supply data to a database column specifically designed to accept unlimited quantities of data, or unrestricted data values, are excepted.

# Suggestion for Priority Level (A/AA/AAA)
Level A (because this is a basic concept in data input programming that should always be followed as a matter of course, though frequently omitted in practice).
# Related Glossary additions or changes
None
# What Principle and Guideline the SC falls within.
Principle 3, Guideline 3.3.
# Description
[This has been written to make it suitable for use in the Understanding document.]

The intent of this success criterion is to ensure authors utilise any mechanisms provided by the technology being used to ensure that form input fields are assigned limits to prevent them accepting input that cannot be handled by subsequent processing (including, but not limited to, saving to a database). 

Most text input systems allow the author to specify the maximum number of characters the user is allowed to enter in text fields. (For example in HTML a text field does not accept additional characters beyond the maximum specified. Most commercial developer frameworks provide a similar limit.)

Most systems with numeric input fields have a means for the author to specify the maximum and minimum numeric values the user is allowed to enter. (Again using HTML5 as an example of this, though it is not quite as robust as refusing further typing, as text fields do; the user can enter a value outside the max/min range, but the field then shows an error when the user leaves it.)

When the author uses these limiting mechanisms, users are made aware, at the earliest possible moment for instance while they are still typing, when they are trying to go beyond the limits that will be accepted. They can adjust their entry immediately, which is far better than making them backtrack and amend it either once they have left the field or later on form submission – a sequence particularly onerous for some groups of disabled people. 

Such limiting mechanisms also avoid the risk of over-long text or out of range data getting through the form completion and submission process, resulting in the user thinking it has been accepted and saved, when in fact the data has been shortened or otherwise corrupted on saving to a database, very likely without reporting back to the user.

Note 1: Sometimes a long string of text, that is saved to a database completely, may be displayed on screen later in shortened form, just a few characters of it for purposes of identifying the particular input while saving space on the screen. This deliberate abbreviation should not be interpreted as falling within the reference in the success criterion to what “can be handled by subsequent processing”.

Note 2: Some database systems allow an open-ended type of input field where there is no limit, in the database system, on the amount of text or other input that can be added. (There may be practical limits in the amount of data that can be stored on the physical hardware holding the database, but that is not an inherent limit of the database.) For such input fields, and assuming no other limit is set in the subsequent processing of data, no limit need be supplied by the author – this is the intent of the exception in the criterion. 

Note 3: The two most widely occurring situations involve the maximum length of text input, and maximum and minimum values in numeric input. However the principle embodied in this success criterion can usefully be applied in any data input method where limits of some kind exist and can be programmed for in the technology used  for input.

Note 4: Many screen readers and similar assistive devices have a noticeable deficiency here. Even on fields that have the maximum length set, when the user reaches the limit the reader makes no announcement of the fact even though a sighted person would be able to see that input is no longer being accepted. This success criterion is not able to address that problem. It is a deficiency of those devices and must be addressed by them in order to provide their users with an improved experience (and one that complies with the spirit of this success criterion).

Note 5: It is understood that not all technologies provide the necessary means to implement this criterion, and it is not intended  to discourage existing input validation methods, but to work in parallel with them. For instance HTML4 and XHTML do not provide max and min attributes for numeric fields; they were added in HTML5. The intention is to provide immediate notice to the user that they are exceeding a limit, where the technology permits it. In time systems will improve and incorporate these mechanisms where they do not already have them.

Note 6: It is felt that the minimum probably need not be set if the minimum is zero. Most numeric input is of positive numbers where a minimum of zero would be assumed unless otherwise stated. For entry of negative numbers the minimum negative number would need to be specified.

# Benefits
This criterion avoids two situations that often occur in online forms:

1)	A user enters invalid data only to be told subsequently, either on leaving the field or when the complete form is submitted, that they are over-limit and must go back and amend their input. In some cases they may have typed a considerable amount of input to no purpose.
2)	Or worse, their data is accepted at the input validation stage (due to weakness in the validation) but subsequently changed behind the scenes (and probably unknown to the user) when it is saved to a database. Some databases have an automatic mechanism, or can optionally be set to implement one, such that data outside the limit is automatically truncated or adjusted in value, on saving, without raising an error condition. 

In the second case the first the user might know about the change to their data is the next time they look at their data in the system – if they notice then - or, more seriously in the case of data used for financial or legal purposes, when some financial or legal problem results from the incorrect data in the future.

## Users who will benefit 

While all users find these situations annoying and frustrating, a number of groups of disabled users can find it more difficult to go back to amend things than non-disabled users. 

These include blind screen and Braille reader users, users of switches, sip and puff devices, and other similar devices for whom form input is difficult enough anyway. Also users with manual dexterity problems regardless of the input device they use. Where data is automatically amended by the system some users with cognitive or memory impairment may not realise it. 

### Example 1: 
The administrators of a blog site accepting guest posts notice that the titles of some article submissions they receive are cut off mid-word. On investigating they discover that, while the title field in the database is only 50 characters long, the title field on the online article entry form has no limit placed on it. Some authors are entering titles without knowing they have exceeded the space limit in the database. So the developers amend the web page to place a 50 character limit on that form field, the same as in the database. And to offer further guidance to users, they include an instruction in the field’s label, making it, “Article title (maximum 50 characters)”.

### Example 2: 
A Last Name field on a website input form has been set to accept a specific maximum number of characters that is expected to adequately cater for all likely surnames in the language and locality the website is written in. A person with an unusually long last name enters it and it appears to be accepted in full. But in the subsequent processing the name is truncated when it is saved to the database. A person in this situation should have been told, during input, about the problem. They would then be able to decide how best to enter their name; they might for instance put a part of it into the First Names field. But they should be the one to make that decision, rather than it simply being truncated by the database save, resulting in possible legal or other complications later.

### Example 3: 
An insurance company arranges travel insurance for people between the ages of 25 and 65. Its online quotation system asks a long list of complex questions about health and status, starting with basic details including age. But the designer and author do not set limits on the age field because in their personal view the explanation of all the various restrictions has already been sufficiently explained elsewhere on the system. The validation of the data is done on form submission when all questions have been completed. So people wanting quotes, and who either missed or forgot the explanations on other pages of the system, frequently fill out all the complex fields only to be told, at the end, that they are not eligible due to their age. They could have found this out at the start, and be saved the trouble, if the author had just set appropriate maximum and minimum limits on numeric fields.

# Testability
### Initial quick check:
Check the source code to confirm that: 
1.	all text fields have a maximum length attribute, and that these lengths look sensible for the fields concerned. 
2.	All numeric input fields have a maximum value attribute, and that these lengths look sensible for the fields concerned.
3.	All numeric input fields have a minimum value attribute except where their value can correctly be zero, and that these lengths look sensible for the fields concerned.

These initial checks will often reveal deficiencies, such as no limits specified, without need for further testing. Another common error by developers is to supply a maximum length, but set it to a generic value (such as 255) on all text fields (they may do this because they do not have the database schema available at the time and forget to correct it later). 

### If fields pass the initial quick checks then:

For each field on the page or document:
1.	In each text field enter as much text as the system allows to be typed up to the end of the field, using a particular character as the end character of each field to identify the end of the input.
2.	In each numeric field enter the maximum possible value the form permits.

When all fields are completed:
3.	Submit the form.
4.	Retrieve the data in whatever way the system provides (or check the database directly).
5.	Check each text field contains the identifying end character to prove nothing has been truncated.
6.	Check each numeric field contains the exact value originally entered into it.
7.	Repeat steps 2 and 6 above but enter the minimum possible value that the form permits.

Fields where unlimited length of input, or unlimited numeric values, are permitted by the database or other processing systems need not be tested.

### Expected Result: 
Steps 5 and 6 will be true - the end character will show in every text field, and the correct value in every numeric field.

Note i: If during Steps 1 and 2 a field is found to allow an excessively longer input than would be reasonable for that data item, seemingly having no limit set, it is probable that the field length has not been set at all (checking the source code may confirm that); in which case that field has failed this test.

Note ii: Fields that are neither simple text nor numeric can be similarly tested, where feasible, using whatever test inputs are relevant to the type of field. If such a test can be run and a failure can be demonstrated, the failure should be reported as a failure of this success criterion.

Note iii: Automated test systems can alert when no limited is specified. When run against HTML or XHTML systems they should still report lack of max and min attributes to alert the authors of the need to move to HTML5. Some automated systems may also be able to input test data in line with step 1 and display the entered and expected data after form submission.

# Techniques
The following would be suitable as General techniques and/or as HTML techniques:

New technique: Specifying the maximum number of characters for a text input field
New technique: Specifying the maximum and minimum values for a numeric input field

**Comment by patrickhlauke** [comment]
This feels more like a usability, rather than accessibility, issue to me. It also annoys/affects all types of users.

**Comment by guyhickling** [comment]
@patrickhlauke Thank you for looking through it, and you have a fair point. However the same could probably be said for a few SCs. For instance the proposed 1.4.10 Zoom Content is to avoid users having to scroll long lines. That is a usability issue for all users, but also much more tiresome for many disabled users.

In forms, a person with hand tremors, or a low vision user who can barely even see the form, finally hits the submit button only to be told they must go back and edit a long line of text into something smaller. It isn't impossible for them, just much more effort than it would be for a non-disabled user. So I think there is a good case for an SC here.

​The good news is that t​he overhead this SC would place on developers is trivial. Using ​maxlength ​(in HTML​ for instance​) to ensure text fields cannot take more characters than the database column ​it feeds into ​can accept is​ ​elementary ​programming. So it is only asking ​some ​developers to do what they should be doing anyway. An SC would simply enable us to pick up on it when they don't!

**Comment by awkawk** [comment]
@guyhickling marking as deferred as we aren't adding additional SC at this time (we closed on new SC proposals August 22).

**Comment by guyhickling** [comment]
@awkawk Will deferred SCs automatically go forward to any v2.2 there may be, or failing that Silver, or will they have to be raised again as fresh submissions?
